Feature: Merge/Link Tickets

aydreeihn opened the pull request:
This new feature adds the ability to merge or link tickets. Please see the link below for the full documentation:

https://github.com/osTicket/docs/blob/fc00e5fa8e34c5c1a868a18ad2662da5df8415dc/Features/Ticket%20Merge%20%26%20Link.rst

I'm open to feedback and suggestions!

Also, please note, this feature requires an upgrade, so do not merge this to your production codebase/database. You should test the feature on a new branch and test database until it is officially merged to the main codebase.

nfebuary committed:
Add preview option of threads in a ticket.

nfebuary committed:
Thread preview design and function modification

Changes to the new ticket thread preview design. Removes avatars and
makes color changes to the header ticket when collapsed. Thread preview
expands when clicked and collapses when header is clicked.

nfebuary committed:
Jquery double click issue fix

Fixes the funky double click after a message is expanded now there are
only single clicks.

protich committed:
Implement Thread Preview

aydreeihn committed:
Ticket Merge Version 1

This feature implements a basic, visual Ticket Merge where thread entries from other Tickets can be made visible on a Parent Ticket.

aydreeihn committed:
Code Fixes

- Do not check merge type for Tasks

aydreeihn committed:
Issue: Ticket # Containing Letters

This commit fixes an issue where we could not merge Tickets if the Ticket# included letters.

aydreeihn committed:
Completely Hide Disabled Fields

If a Field is disabled, it will not be shown anywhere in the system (ticket view, search, export).

aydreeihn committed:
Code Fixes:

- Ensure that we do not put the merged icon on task threads
- Make sure we remove the parent_id on child tickets if we delete the parent ticket

aydreeihn committed:
Additional PHP 7.1 fixes for class.filter.php

Commit:
PHP 7.1 Errors

This commit fixes PHP 7.1 warning messages.

JediKev committed:
merge: Set Children Statuses

This pull adds the option to set the status of all Child Tickets if setting
status of a Parent Ticket. This is helpful for situations like closing a
Parent Ticket and needing to close all Children as well. This option is only
visible when setting the status of a Ticket with Children.

aydreeihn committed:
Merge Thread Entries Fix

This commit fixes an issue on the order of merged thread entries. We needed to do a UNION instead of a UNION ALL in order for them to show up in the correct order.

aydreeihn committed:
Option to hide Children Thread Entries

This commit gives Agents the option to hide children thread entries on parent Tickets using a checkbox on the Merge Tickets modal. If hidden, the Agent is still able to see the related children Tickets in the 'Related Tickets' tab of the Parent Ticket. The thread entries will be hidden by default.

JediKev committed:
issue: getObjectType Typo

This addresses a typo of `getObJectType()`, this should be
`getObjectType()`.

aydreeihn committed:
Ticket Merge Feature Additions

- Change visual flag to separate thread flag. Visual should be the default since messages will route to them the same way they route to unmerged tickets
- Modify code as needed to account for possibly not having a thread when getThread() is called
- UI change to merge ticket template to include Merge Types
- Allow visually merged tickets to be split
- Make thread entries display according to the merge types for Combined and Separate merges
- Make responses to child tickets for combined or separate merges thread to the parent ticket
    - When posting a message, we need to check to see if the ticket is a parent or a child, and if it is for a child, which type of merge it is for. From there, we must decide which ticket the message should go to and route it accordingly with the new $ticket variable (instead of using $this).
    - Block replies on child tickets
    - Create an object type 'C' for child tickets that will allow us to correctly set the extra field for incoming thread entries that were made to children tickets

aydreeihn committed:
Make sure we add thread events for merge/split in both the parent and child ticket.

aydreeihn committed:
For merged threads, when hovering over the merge icon, put the Ticket # if possible, otherwise, put 'Merged'

aydreeihn committed:
Issue: Ticket View Permissions for Merged Tickets

This commit changes merged Ticket permissions to the following:

If the Agent does not have access to the Child ticket and the threads are merged, they should not be allowed to view the Parent. If the threads are not merged, the Agent should still be able to access the Parent Ticket.

aydreeihn committed:
Merge Availability

Make sure we do not show the rest of the form to add more tickets if merge is already in place

aydreeihn committed:
Delete Child Ticket Option

This commit adds an option for the Agent to delete the child ticket when doing a Separarate or Combined merge. A warning pops up so the Agent knows which Ticket will be deleted.

aydreeihn committed:
Allow Further Merging

- make it to where you can merge other tickets to a parent, but the options are locked down for already merged tickets.
	Ex: if already separate option, that should be the only option moving forward
- prevents admins from trying to merge tickets together if more than one parent ticket is selected for merging.
- make sure only visual merges allow you to change who is the parent

aydreeihn committed:
Merged Thread Fixes

- Do not show collaborator information for merged tickets because their thread is now the parent ticket's
- only do the thread lookup one time to eliminate unnecessary calls
- Adjust getThread and getThreadId methods to account for the merge feature
	- getThread adjustments allow us to still see the thread events for child tickets

aydreeihn committed:
Delete Child Ticket Issue

This commit ensures that if you choose to delete child tickets when doing a merge, we set the parent ticket's merge type to visual. Since the child ticket merged into it no longer exists, the parent ticket should act as a normal ticket that can continue to be merged as needed with any other ticket.

aydreeihn committed:
Merge Indicators

This commit makes sure that we show merge indicators, whether it is an annotation or the Merged column, for both parent and child tickets.

aydreeihn committed:
Disable Children From Merge

This commit disables children from being allowed to be a part of new merges on mass process merge

aydreeihn committed:
Object Ensurance

Check to see if we have Threads/Collaborators before trying to call a function from them.

aydreeihn committed:
Disable Children From Merge Part 2

This commit disables children from being allowed to be a part of new merges when merging from an individual ticket. It also makes sure that the merge type cannot be changed.

aydreeihn committed:
Change 'Visual Merge' to Link

- Create Role permission for linking tickets
- Logic to show/hide merge or link option depending on what is chosen
- Change 'split' event to 'unlink'
- Write thread events for link and unlink
- Add link annotation
- Add link column
- Make merge and link available in advanced searches
- Make merge and link exportable
- Make sure merged Tickets cannot be linked

aydreeihn committed:
Option to Move User/Collaborators

When merging tickets, give the option to move the user and/or collaborators if they differ from the current parent user.

aydreeihn committed:
Code Fixes/Cleanup

- Set the linked flag in the setMergeType function
- Make sure the parent and children tickets' merge types are set each time merge is called
- Make sure we always log a thread event for merges/links on all related tickets
- Make sure no flags are set if child tickets are deleted upon merge
- Make sure we save the parent ticket when choosing whether to show child tickets in link
- Fix the reload page function

aydreeihn committed:
Merging a Linked Parent

If a linked parent is merged into another ticket w/a separate or combine merge, what should happen to the linlked children?
	I want it to just unlink the children and handle the parent as usual

aydreeihn committed:
Remove Show Children from Link

This commit removes the option to show child thread entries on visual merges. If an Agent wants to see all thread entries together, they should perform a merge. If they want to see thread entries from child tickets somewhere within the Parent Ticket, they should use Thread Preview on the Related Tickets tab.

aydreeihn committed:
Feature Cleanup/Modifications

- Put checkboxes before label text on Merge/Link form
- Move 'Delete Child Ticket' option to bottom of form
- Add extra info to tickets on Merge/Link list
	- Subject
- Minor fix with Link annotation so it shows up everywhere
- Make the tickets div in the Merge/Link form scrollable

aydreeihn committed:
Front End Merge Handling

- Show merged thread events to Users
- Put a warning for the Users logged into the front end to let them know they can’t reply to merged children
        - Put a link to the Parent where they can reply
- Allow a User to view and reply to a parent ticket even if they're not the ticket owner or a collaborator (they can reply via email anyway)
	- This will add them as a collaborator to the parent
- During a merge, if the users are differnt, give the option to move users or users + collaborators to the parent (remove the 'none' option)

aydreeihn committed:
Restructure Code

- Make sure that for both mass process and ticket-view merge/link, we're getting tickets by filtering,not lookups
- Move getMergeType beside relevant functions
- Create function to get merge types based on flags
- Add Subject and User Name to Merge/Link form

aydreeihn committed:
Task Handling

- Move tasks to parent if deleting children
- Make sure thread entries work correctly for tasks and tickets

aydreeihn committed:
Code Fixes/Cleanup

- make sure ticket #, subject, and user name appear when adding new tickets to the form
- use getThread() function instead of ->thread since a ticket could be a child which has no thread (it uses the parent thread)

aydreeihn committed:
Task Handling

- Give an option to move child tasks to parents for merges
- Add a task count to the merge/link field
- Make sure the tickets display the same info on mass process and adding tickets individually

aydreeihn committed:
Merge/Link Annotation + Enhance Ticket Listing

This commit adds an icon to the merge tickets form that shows child tickets when hovered over.

Additionally:
- pull annotations to the far right
- Only show annotations if the count is > 0
- Decide whether to show link or merge annotation based on the merge type
- Add a new collaborator annotation for the collaborator preview
- Put the ticket preview on the ticket number rather than collaborator preview

aydreeihn committed:
Participant Options

This commit will give the Agent the option to move participants to the parent ticket based on whether the user on multiple tickets is the same or different as well as if there are any collaborators involved in the tickets.

aydreeihn committed:
Thread Entry Annotation + Code Cleanup

- Add annotation for thread entry count (if any)
- Fix ticket links

aydreeihn committed:
Ticket Merge Patch

- Update event.yaml
- Events added to newest patch

Note: new fields were already done previously w/merge version 1

aydreeihn committed:
Code Cleanup

- make user + collabs the default option. make sure they still have option to drop collabs if the users are the same
- Space out the 'Delete Child Ticket' and 'Move Child Tasks to Parent' checkboxes
- We should show the Participants dropdown regardless of users on Tickets because if it was ever hidden, other Tickets could be added to the form that would need to see the options and it's a static form field (doesn't pop up if a Ticket is added w/diff user or collabs)

aydreeihn committed:
Hide Manage Collaborators for Merge/Link

- Pass a new variable, $manage, to the previewCollaborators function (and in turn the showCollaborators function) so that we can specify when we want the 'Manage Collaborators' option to be available
- Set the new variable every time one of those functions are called

aydreeihn committed:
Referrals and Child Relationships

- Create a child_thread relationship on tickets so that children tickets can be searched
- For Ticket::getThread if the ticket does not have a thread but it has a child_thread, return the child_thread
- Add referred children tickets to staff's ticket visibility
- Do not change the object_id in the thread table
- Allow ticket merge if Agent has referral access to a Ticket
- Hide children tickets from queues
- Make sure children tickets can be searched
- Remove getIdByObjectId function (no longer needed)
- Remove children tickets from queue counts

aydreeihn committed:
Merge/Link Annotation Alignment

Put annotations in line w/one another on the merge/link modal.

aydreeihn committed:
Automatic Merge Referrals

When merging tickets, if children have a different department from the parent, refer the children departments to the parent ticket. Otherwise, Agents would only be able to access the children ticket they used to have access to with a link or search because we hide merged children from the ticket queue.

aydreeihn committed:
Linked Icon - Searches

Make sure we still show the linked icon if necessary on searches. Merge and Link follow the same criteria but look for different flags, so if the merge flag isn't on a ticket, we should then check for the link flag

aydreeihn committed:
Code Cleanup

- Reduce lines of code in 'Related Tickets' tab
- Wording tweaks
- Restructure unlinking
- Make thread entry and collaborator count distinct
- Make sure all merged tickets (children included) show up for searches and saved searches
- Make sure the count for saved searches will count children merges
- Clean up redundancy in general
- Fix linking sort order (if set)
- Make sure child thread entries are separated and then sorted for separate merges

protich committed:
Field Visibility: Add multi-option constraint

This commit adds ability to add visibility constraint that is dependent on
multiple options. It also addresses cases where constrain is on an existing
or misspelled field.

protich committed:
CheckboxWidget: Support setting value by field Id.

This commit adds ability to set initial checkbox value via source data

protich committed:
Form::process

Adds a routine similar to getClean but forms downstream can use it to return
database ready data. This is especially necessary when UI elements don't
match database fields.

protich committed:
CheckboxWidget: getValue

Return default value when data source is not set - this is necessary for
Visibility Constraint that's dependent on the value of a checkbox field.

protich committed:
ChoicesWidget: Allow numeric zero option

Use strict type check comparison when checking if a choice value equals the
default value. This is necessary to allow for choices with numeric zero (0)
    as the value.

protich committed:
Datetime: Use Format::parseDateTime

protich committed:
Datetime: Use Format::parseDateTime

protich committed:
Datetime: Use Format::parseDateTime

protich committed:
TimeField: Add initial stump of TimeField

This commits introduces TimeField and it's picker widget. The field uses
jQuery ui & datetimepicker addon for time input.

protich committed:
VisibilityConstraint: Change watcher

Make JS checker function unique for each field/widget.

aydreeihn committed:
Code Cleanup

- Remove unneeded code
- Reduce lines of code
- Fix line spacing

aydreeihn committed:
Switch Merge Type

This commit allows you to change the merge type between combine and separate for merged tickets within the parent ticket

aydreeihn committed:
Display Trash Icon

Show the trash icon in the merge/link form as long as the tickets selected are not part of a merge

aydreeihn committed:
Move entries of deleted children

If children tickets are deleted and then the parent is merged to another ticket and becomes a child, we have to move all thread entries, even if they were child entries.

aydreeihn committed:
Child Thread on Parent Delete

If a parent ticket is deleted, we need to go to each child thread and set the object type back to 'T' so that those tickets will show up in the ticket queue again.

Note: If a message is emailed to the child tickets (now regular tickets), they thread back to that ticket as normal

aydreeihn committed:
Minor Fixes

- allow changing merge type on mass process as well
- fix error message for merged chidlren being selected (add ticket_pid back)
- add an error to not allow merged tickets to be added manually to the form. want them to be checked in mass process so they will sort correctly
- show children tickets in the merge form within ticket view
- fix vertical scroll on merge/link modal
- scoot the merge/link icons div to the left so they don't run into the icon-trash button

aydreeihn committed:
When to Unlink

- Fix logic for when we should unlink tickets (changing from link to merge or changing link parent)
- Make sure if we're unlinking to change the parent we still log the link event for the new link

aydreeihn committed:
Ticket Preview Info

- Add the merge/link icon to the ticket preview

aydreeihn committed:
Ticket Print Sort Order

This commit makes sure that the thread entries on printed tickets match the same order as the ticket in the helpdesk. A new function, sortEntries was added to class.thread.php so it can be called for the entries template as well as the ticket print template.

We also make sure that the order of thread entries display correctly for the client side tickets and printed pdfs of tickets.

aydreeihn committed:
Ticket Print Sort Order

This commit makes sure that the thread entries on printed tickets match the same order as the ticket in the helpdesk. A new function, sortEntries was added to class.thread.php so it can be called for the entries template as well as the ticket print template.

We also make sure that the order of thread entries display correctly for the client side tickets and printed pdfs of tickets.

protich committed:
Add ORDINAL suffix to a number

e.g 1st, 2nd, 3rd etc.

protich committed:
Add setting to hide/show timezone hint

This is necessary for timezone agnostic date/time inputs.

protich committed:
FormRenderer: Show empty Label

This is necessary to make sure fields still line-up even when label is not
provided.

protich committed:
VisibilityConstraint: Add support for neq

Support 'not equal' on visibility constraints

protich committed:
Introducing the concept of Schedules

Schedules can be used to specify when SLA are active by defining Business
Hours and Holidays.

protich committed:
CheckOverdue: ORM implementation

This commit implements checking for overdue tickets in ORM.

protich committed:
Onetime Entries: Stop on end date & time

protich committed:
Fixes: Address PHP 7.3 warnings

Also makes sure schedule is truly saved in sla plan page

All Credit goes to @aydreeihn

aydreeihn committed:
Modify Upgrader

This feature has been updated to be the patch after Schedules (#4984). That being said, this feature requires an upgrade as well, so please do not merge this to your production database. You should test the feature on a new branch until it is officially merged to the main codebase.

protich commented on include/upgrader/streams/core/8b923d61-9b5550da.patch.sql:
```diff
@@ -1,5 +1,5 @@
 /**
-* @signature 49fb8af84d9ac386a573f3eb80d8a0ee
+* @signature 9b5550dad8d3f036d62eb1340f230317
```
Change @version on this file to 1.14

aydreeihn replied:
Done!

ntozier commented:
@aydreeihn you are my favorite person right now. :)

protich committed:
Make sure Schedule Entries are unique per Schedule

The system assumes one entry per day - meaning no individual day chunking is
allowed within the same schedule unless it's an exclusion via a Holiday
schedule on a Business Hours schedule.

protich committed:
oopsie: Revert mistaken change on field__op split

Commit af322a41d introduces a new field operator and in an attempt to
simplify how a field and it's operator is split a bug was introduced when a
field with a nested complex name is being split.

This commit reverts to the original "complex" splitter.

philbertphotos commented:
This is life now.

Mike-Benoit commented:
Finally! This will be amazing and it has a ton of interesting functionality, great job! 

However after reading the full documentation, I didn't see anything about splitting existing tickets? In our case we would use this probably almost as much as linking/merging tickets and it seems like the two features would be quite similar to implement together.

aydreeihn commented:
@Mike-Benoit What exactly do you mean by splitting? We do have the option to create a ticket from a thread. So for example, if an Agent felt it might be better to reply to a User's message in a separate ticket/conversation, they could create a new ticket where that thread would be the first message of a new ticket.

![Screen Shot 2019-08-29 at 11 12 00 AM](https://user-images.githubusercontent.com/22985133/63957486-18921f00-ca4e-11e9-910d-4aa96037b5b2.png)

![Screen Shot 2019-08-29 at 11 12 25 AM](https://user-images.githubusercontent.com/22985133/63957499-1def6980-ca4e-11e9-9a9c-af797009061b.png)

![Screen Shot 2019-08-29 at 11 12 41 AM](https://user-images.githubusercontent.com/22985133/63957514-22b41d80-ca4e-11e9-9231-2d1003791177.png)

![Screen Shot 2019-08-29 at 11 12 56 AM](https://user-images.githubusercontent.com/22985133/63957528-2778d180-ca4e-11e9-8483-230cd5375072.png)

TechCare-Aus commented:
@aydreeihn this is going to save me SOOO much time. So thank you very much. Just curious as to why the release date is pending v1.14 and we are on v1.12 :(

JediKev commented:
@TechCare-Aus

`v1.14` will be our next major realease. We are skipping `v1.13` as it’s an unlucky/bad number; like you don’t see hotels with a 13th floor. (triskaidekaphobia)

Cheers.

philbertphotos commented:
This good enough for production in a 1.12 environment?

f1ynng8 commented:
Any rough plan to release v1.14?

aydreeihn committed:
Ticket Print Thread Event Times

This commit ensures that if we print thread events on Ticket PDFs, the time is never relative time, even if the agent has their time set to relative. Relative time should only be an option when viewing a Ticket online.

Additionally, we do not need to put anything related to relative time in the ticket-print template.

Mike-Benoit commented:
> @Mike-Benoit What exactly do you mean by splitting? We do have the option to create a ticket from a thread. So for example, if an Agent felt it might be better to reply to a User's message in a separate ticket/conversation, they could create a new ticket where that thread would be the first message of a new ticket.
> 
> ![Screen Shot 2019-08-29 at 11 12 00 AM](https://user-images.githubusercontent.com/22985133/63957486-18921f00-ca4e-11e9-910d-4aa96037b5b2.png)
> 
> ![Screen Shot 2019-08-29 at 11 12 25 AM](https://user-images.githubusercontent.com/22985133/63957499-1def6980-ca4e-11e9-9a9c-af797009061b.png)
> 
> ![Screen Shot 2019-08-29 at 11 12 41 AM](https://user-images.githubusercontent.com/22985133/63957514-22b41d80-ca4e-11e9-9231-2d1003791177.png)
> 
> ![Screen Shot 2019-08-29 at 11 12 56 AM](https://user-images.githubusercontent.com/22985133/63957528-2778d180-ca4e-11e9-8483-230cd5375072.png)

Creating a new ticket from a thread appears like it might help in some of the cases. The situation that seems to occur (way too often) for us is the customer creates the original ticket regarding the topic of "Apples", then replies to that same ticket for a question related to "Oranges". Its not always obvious right away that the two aren't related, so sometimes we could be 3+ exchanges in before we figure this out.

So ideally we would have a way to select X thread entries and move them to a completely new ticket, or even merge them with some other existing ticket. The PHPBB forum software has a really nice moderation tool for doing exactly that, but essentially its just checkboxes that allow you to select which thread entries you want to move, and if you want a pointer to the new topic (ticket) or not. 

Either way I look forward to giving this feature a try!

neoand commented:
> @aydreeihn you are my favorite person right now. :)

Me too !!!!!
